File: Algorithm/Code_Materials/c_or_cpp/charpter8_tree_hot_problems/level2/max_depth.cpp
#include <vector>
#include <stack>
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int maxDepth(TreeNode *root)
{
    if (root == nullptr)
        return 0;
    return max(maxDepth(root->left), maxDepth(root->right)) + 1;
}

int maxDepth2(TreeNode *root)
{
    if (root == nullptr)
        return 0;
    queue<TreeNode *> Q;
    Q.push(root);
    int ans = 0;
    while (!Q.empty())
    {
        int sz = Q.size();
        while (sz > 0)
        {
            TreeNode *node = Q.front();
            Q.pop();
            if (node->left)
                Q.push(node->left);
            if (node->right)
                Q.push(node->right);
            sz -= 1;
        }
        ans += 1;
    }
    return ans;
}

int main()
{
    // 创建测试用例的二叉树
    TreeNode *root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(2);
    root1->left->left = new TreeNode(3);
    root1->left->right = new TreeNode(4);
    root1->right->left = new TreeNode(4);
    root1->right->right = new TreeNode(3);

    cout << maxDepth2(root1) << endl;
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/Java/src/main/java/com/yugutou/charpter8_tree_hot_problems/level2/MaxDepth.java
package com.yugutou.charpter8_tree_hot_problems.level2;

import com.yugutou.tools.BinaryTree;
import com.yugutou.tools.TreeNode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * leetcode104 最大深度和最小深度问题
 */
public class MaxDepth {
    public static void main(String[] args) {
        BinaryTree bTree = new BinaryTree();
        bTree.root = bTree.buildBinaryTree();
        bTree.root = bTree.buildBinaryTreeComplex();
        int testMethod = 3;
        int result = 0;
        switch (testMethod) {
            case 1://通过递归计算二叉树的深度
                result = maxDepth_1(bTree.root);
                break;
            case 2://通过层次遍历计算二叉树的深度
                result = maxDepth_2(bTree.root);
                break;


        }

        System.out.println(result);
    }

    /**
     * 通过递归计算二叉树的深度
     *
     * @param root
     * @return
     */
    public static int maxDepth_1(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftHeight = maxDepth_1(root.left);
        int rightHeight = maxDepth_1(root.right);
        return Math.max(leftHeight, rightHeight) + 1;

    }

    /**
     * 通过层次遍历来求最大深度
     *
     * @param root
     * @return
     */
    public static int maxDepth_2(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        int ans = 0;
        while (!queue.isEmpty()) {
            //size表示某一层的所有元素数
            int size = queue.size();
            //size=0 表示一层访问完了
            while (size > 0) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
                size--;
            }
            ans++;
        }
        return ans;
    }


}
